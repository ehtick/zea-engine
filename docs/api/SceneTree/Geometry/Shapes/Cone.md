<a name="Cone"></a>

### Cone 
Represents a cone geometry.

```
const cone = new Cone(1.2, 4.0)
```

**Parameters**
* **Radius([`NumberParameter`](api/SceneTree\Parameters\NumberParameter.md)):** Specifies the radius of the base of the cone.
* **Height([`NumberParameter`](api/SceneTree\Parameters\NumberParameter.md)):** Specifies the height of the cone.
* **Detail([`NumberParameter`](api/SceneTree\Parameters\NumberParameter.md)):** Specifies the number of subdivisions around the `Z` axis.
* **Cap([`BooleanParameter`](api/SceneTree\Parameters\BooleanParameter.md)):** Specifies whether the base of the cone is capped or open.


**Extends**: <code>[ProceduralMesh](api/SceneTree\Geometry\Shapes\ProceduralMesh.md)</code>  
<a name="new_Cone_new"></a>

### new Cone
Create a cone.


| Param | Type | Default | Description |
| --- | --- | --- | --- |
| radius | <code>number</code> | <code>0.5</code> | The radius of the base of the cone. |
| height | <code>number</code> | <code>1</code> | The height of the cone. |
| detail | <code>number</code> | <code>32</code> | The detail of the cone. |
| cap | <code>boolean</code> | <code>true</code> | A boolean indicating whether the base of the cone is capped or open. |



### [Class Tests](api/SceneTree\Geometry\Shapes/Cone.test)