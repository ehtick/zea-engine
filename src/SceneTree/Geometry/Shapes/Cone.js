import { Vec2 } from '../../../Math/Vec2';
import { Vec3 } from '../../../Math/Vec3';
import { Mesh } from '../Mesh.js';
import {
  BooleanParameter,
  NumberParameter
} from '../../Parameters';
import {
  sgFactory
} from '../../SGFactory.js';

class Cone extends Mesh {
  constructor(radius = 0.5, height = 1.0, detail = 32, cap = true) {
    super();

    if(isNaN(radius) || isNaN(height) || isNaN(detail))
      throw("Invalid geom args");

    this.__radiusParam = this.addParameter(new NumberParameter('radius', radius));
    this.__heightParam = this.addParameter(new NumberParameter('height', height));
    this.__detailParam = this.addParameter(new NumberParameter('detail', ((detail >= 3) ? detail : 3), [3, 200], 1));
    this.__capParam = this.addParameter(new BooleanParameter('cap', cap));

    this.addVertexAttribute('texCoords', Vec2);
    this.addVertexAttribute('normals', Vec3);
    this.__rebuild();

    const resize = ()=>{
      this.__resize();
    }
    const rebuild = ()=>{
      this.__rebuild();
    }
    this.__radiusParam.valueChanged.connect(resize); 
    this.__heightParam.valueChanged.connect(resize); 
    this.__detailParam.valueChanged.connect(rebuild); 
    this.__capParam.valueChanged.connect(rebuild); 
  }

  get radius() {
    return this.__radiusParam.getValue();
  }

  set radius(val) {
    this.__radiusParam.setValue(val);
    this.__resize();
  }

  get height() {
    return this.__heightParam.getValue();
  }

  set height(val) {
    this.__heightParam.setValue(val);
    this.__resize();
  }

  get detail() {
    return this.__detailParam.getValue()
  }

  set detail(val) {
    this.__detailParam.setValue(val);
    this.__rebuild();
  }


  get cap() {
    return this.__capParam.getValue();
  }

  set cap(val) {
    this.__capParam.setValue(val);
    this.__rebuild();
  }

  __rebuild() {

    this.clear()

    const nbSides = this.__detailParam.getValue();
    const radius = this.__radiusParam.getValue();
    const height = this.__heightParam.getValue();
    const cap = this.__capParam.getValue();
    let numVertices = nbSides + 1;
    if (cap) {
      numVertices += 1;
    }
    this.setNumVertices(numVertices);
    const tipPoint = nbSides;
    const basePoint = nbSides + 1;

    //////////////////////////////
    // Set Vertex Positions

    this.getVertex(tipPoint).set(0.0, 0.0, height);
    for (let i = 0; i < nbSides; i++) {
      let theta = (i / nbSides) * 2.0 * Math.PI;
      this.getVertex(i).set(radius * Math.cos(theta), radius * Math.sin(theta), 0.0);
    }
    if (cap) {
      this.getVertex(basePoint).set(0.0, 0.0, 0.0);
    }

    //////////////////////////////
    // build the topology
    this.setFaceCounts([nbSides + (cap ? nbSides : 0)]);
    for (let i = 0; i < nbSides; i++) {
      let j = (i + 1) % nbSides;
      this.setFaceVertexIndices(i, j, i, tipPoint);
    }
    if (cap) {
      for (let i = 0; i < nbSides; i++) {
        let j = (i + 1) % nbSides;
        this.setFaceVertexIndices(nbSides + i, i, j, basePoint);
      }
    }

    //////////////////////////////
    // setNormals
    let normals = this.getVertexAttribute('normals');

    let normalElevation;
    let divider = height;
    if (Math.abs(height) < 1.0e-12)
      normalElevation = height < 0 ? -1.0e-12 : 1.0e-12;
    normalElevation = radius / divider;

    let tri = 0;
    for (let i = 0; i < nbSides; i++) {
      let theta1 = ((i + 1) / nbSides) * 2.0 * Math.PI;
      let theta2 = (i / nbSides) * 2.0 * Math.PI;
      let theta = (theta1 + theta2) * 0.5;

      normals.setFaceVertexValue(tri, 0, new Vec3(Math.cos(theta1), normalElevation, Math.sin(theta1)).normalize());
      normals.setFaceVertexValue(tri, 1, new Vec3(Math.cos(theta2), normalElevation, Math.sin(theta2)).normalize());
      normals.setFaceVertexValue(tri, 2, new Vec3(Math.cos(theta), normalElevation, Math.sin(theta)).normalize());
      tri++;
    }
    if (cap) {
      let normal = new Vec3(0.0, -1.0, 0.0);
      for (let i = 0; i < nbSides; i++) {
        normals.setFaceVertexValue(tri, 0, normal);
        normals.setFaceVertexValue(tri, 1, normal);
        normals.setFaceVertexValue(tri, 2, normal);
        tri++;
      }
    }

    //////////////////////////////
    // setUVs
    let texCoords = this.getVertexAttribute('texCoords');

    // Now set the attrbute values
    tri = 0;
    for (let i = 0; i < nbSides; i++) {
      texCoords.setFaceVertexValue(tri, 0, new Vec2((i + 1) / nbSides, 0.0));
      texCoords.setFaceVertexValue(tri, 1, new Vec2(i / nbSides, 0.0));
      texCoords.setFaceVertexValue(tri, 2, new Vec2((i + 0.5) / nbSides, 1.0));
    }
    if (cap) {
      for (let i = 0; i < nbSides; i++) {
        texCoords.setFaceVertexValue(tri, 0, new Vec2(i / nbSides, 0.0));
        texCoords.setFaceVertexValue(tri, 1, new Vec2((i + 1) / nbSides, 0.0));
        texCoords.setFaceVertexValue(tri, 2, new Vec2((i + 0.5) / nbSides, 1.0));
        tri++;
      }
    }

    this.setBoundingBoxDirty();
  }

  __resize() {
    const nbSides = this.__detailParam.getValue();
    const radius = this.__radiusParam.getValue();
    const height = this.__heightParam.getValue();
    const cap = this.__capParam.getValue();

    const tipPoint = nbSides;
    const basePoint = nbSides + 1;

    this.getVertex(tipPoint).set(0.0, 0.0, height);
    for (let i = 0; i < nbSides; i++) {
      let theta = (i / nbSides) * 2.0 * Math.PI;
      this.getVertex(i).set(radius * Math.cos(theta), radius * Math.sin(theta), 0.0);
    }
    if (this.__cap) {
      this.getVertex(basePoint).set(0.0, 0.0, 0.0);
    }

    this.setBoundingBoxDirty();
  }

};

sgFactory.registerClass('Cone', Cone);
export {
  Cone
};
