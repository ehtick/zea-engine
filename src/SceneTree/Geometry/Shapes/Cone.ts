/* eslint-disable no-unused-vars */
import { Vec2 } from '../../../Math/Vec2'
import { BooleanParameter, NumberParameter } from '../../../SceneTree/Parameters/index'
import { Registry } from '../../../Registry'
import { ProceduralMesh } from './ProceduralMesh'
import { Vec3Attribute } from '../Vec3Attribute'
import { Vec2Attribute } from '../Vec2Attribute'

/**
 * Represents a cone geometry.
 *
 * ```
 * const cone = new Cone(1.2, 4.0)
 * ```
 *
 * **Parameters**
 * * **Radius(`NumberParameter`):** Specifies the radius of the base of the cone.
 * * **Height(`NumberParameter`):** Specifies the height of the cone.
 * * **Detail(`NumberParameter`):** Specifies the number of subdivisions around the `Z` axis.
 * * **Cap(`BooleanParameter`):** Specifies whether the base of the cone is capped or open.
 *
 * @extends {ProceduralMesh}
 */
class Cone extends ProceduralMesh {
  protected __cap!: boolean
  protected __capParam: BooleanParameter
  protected __detailParam: NumberParameter
  protected __heightParam: NumberParameter
  protected __radiusParam: NumberParameter
  // topologyParams: string[]

  /**
   * Create a cone.
   * @param {number} radius - The radius of the base of the cone.
   * @param {number} height - The height of the cone.
   * @param {number} detail - The detail of the cone.
   * @param {boolean} cap -  A boolean indicating whether the base of the cone is capped or open.
   */
  constructor(radius = 0.5, height = 1.0, detail = 32, cap = true) {
    super()
    this.topologyParams = []
    if (isNaN(radius) || isNaN(height) || isNaN(detail)) throw new Error('Invalid geom args')

    this.__radiusParam = this.addParameter(new NumberParameter('Radius', radius)) as NumberParameter
    this.__heightParam = this.addParameter(new NumberParameter('Height', height)) as NumberParameter
    this.__detailParam = this.addParameter(
      new NumberParameter('Detail', detail >= 3 ? detail : 3, [3, 200], 1)
    ) as NumberParameter
    this.__capParam = this.addParameter(new BooleanParameter('Cap', cap)) as BooleanParameter

    this.addVertexAttribute('texCoords', new Vec2Attribute())
    this.addVertexAttribute('normals', new Vec3Attribute())

    this.topologyParams.push('Detail')
    this.topologyParams.push('Cap')
  }

  /**
   * The rebuild method.
   * @private
   */
  rebuild(): void {
    const nbSides = this.__detailParam.getValue() || 32
    const radius = this.__radiusParam.getValue() || 0.5
    const height = this.__heightParam.getValue() || 1.0
    const cap = this.__capParam.getValue()
    let numVertices = nbSides + 1
    if (cap) {
      numVertices += 1
    }
    this.setNumVertices(numVertices)
    const tipPoint = nbSides
    const basePoint = nbSides + 1

    // ////////////////////////////
    // Set Vertex Positions
    const positions = <Vec3Attribute>this.getVertexAttribute('positions')

    if (positions) {
      positions.getValueRef(tipPoint).set(0.0, 0.0, height)
      for (let i = 0; i < nbSides; i++) {
        const theta = -((i / nbSides) * 2.0 * Math.PI)
        positions.getValueRef(i).set(radius * Math.cos(theta), radius * Math.sin(theta), 0.0)
      }
      if (cap) {
        positions.getValueRef(basePoint).set(0.0, 0.0, 0.0)
      }
    }

    // ////////////////////////////
    // Build the topology
    this.setFaceCounts([nbSides + (cap ? nbSides : 0)])
    for (let i = 0; i < nbSides; i++) {
      const j = (i + 1) % nbSides
      this.setFaceVertexIndices(i, [j, i, tipPoint])
    }
    if (cap) {
      for (let i = 0; i < nbSides; i++) {
        const j = (i + 1) % nbSides
        this.setFaceVertexIndices(nbSides + i, [i, j, basePoint])
      }
    }

    // ////////////////////////////
    // setUVs
    const texCoords = <Vec2Attribute>this.getVertexAttribute('texCoords')
    if (texCoords) {
      // Now set the attrbute values
      let tri = 0
      for (let i = 0; i < nbSides; i++) {
        if ('setFaceVertexValue' in texCoords) {
          texCoords.setFaceVertexValue(tri, 0, new Vec2((i + 1) / nbSides, 0.0))
          texCoords.setFaceVertexValue(tri, 1, new Vec2(i / nbSides, 0.0))
          texCoords.setFaceVertexValue(tri, 2, new Vec2((i + 0.5) / nbSides, 1.0))
        }
      }
      if (cap) {
        for (let i = 0; i < nbSides; i++) {
          texCoords.setFaceVertexValue(tri, 0, new Vec2(i / nbSides, 0.0))
          texCoords.setFaceVertexValue(tri, 1, new Vec2((i + 1) / nbSides, 0.0))
          texCoords.setFaceVertexValue(tri, 2, new Vec2((i + 0.5) / nbSides, 1.0))
          tri++
        }
      }
    }

    this.resize()
  }

  /**
   * The resize method.
   * @private
   */
  resize(): void {
    const nbSides = this.__detailParam.getValue() || 32
    const radius = this.__radiusParam.getValue() || 0.5
    const height = this.__heightParam.getValue() || 1.0

    const tipPoint = nbSides
    const basePoint = nbSides + 1

    const positions = <Vec3Attribute>this.getVertexAttribute('positions')
    if (positions) {
      positions.getValueRef(tipPoint).set(0.0, 0.0, height)
      for (let i = 0; i < nbSides; i++) {
        const theta = -((i / nbSides) * 2.0 * Math.PI)
        positions.getValueRef(i).set(radius * Math.cos(theta), radius * Math.sin(theta), 0.0)
      }
      if (this.__cap) {
        positions.getValueRef(basePoint).set(0.0, 0.0, 0.0)
      }
    }

    // Note: this breaks an infinite loop where computeVertexNormals calls update which calls rebuild.
    this.dirtyTopology = false
    this.dirtyVertices = false
    this.computeVertexNormals()
  }
}

Registry.register('Cone', Cone)
export { Cone }
