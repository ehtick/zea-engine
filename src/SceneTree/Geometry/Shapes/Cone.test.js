import { Cone } from './Cone'

describe('Cone', () => {
  it('tests default parameters', () => {
    const cone = new Cone()

    expect(cone.getParameter('Radius').getValue()).toBe(0.5)
    expect(cone.getParameter('Height').getValue()).toBe(1)
    expect(cone.getParameter('Detail').getValue()).toBe(32)
    expect(cone.getParameter('Cap').getValue()).toBe(true)
  })

  it('updates parameters', () => {
    const cone = new Cone()
    cone.getParameter('Radius').setValue(1)
    cone.getParameter('Height').setValue(2)
    cone.getParameter('Detail').setValue(16)
    cone.getParameter('Cap').setValue(false)

    expect(cone.getParameter('Radius').getValue()).toBe(1)
    expect(cone.getParameter('Height').getValue()).toBe(2)
    expect(cone.getParameter('Detail').getValue()).toBe(16)
    expect(cone.getParameter('Cap').getValue()).toBe(false)
  })

  it('saves to JSON (serialization).', () => {
    const cone = new Cone(2, 5, 8, true)
    const outputJSON = cone.toJSON()

    expect(outputJSON).toMatchSnapshot()
  })

  it('restores from JSON (serialization).', () => {
    const cone = new Cone()
    const expectedOutput =
      '{"params":{"Radius":{"value":2},"Height":{"value":5},"Detail":{"value":8,"range":[3,200],"step":1},"Cap":{"value":true}},"type":"Cone","numVertices":10,"vertexAttributes":{"positions":{"data":[2,0,0,1.4142135381698608,1.4142135381698608,0,1.2246468525851679e-16,2,0,-1.4142135381698608,1.4142135381698608,0,-2,2.4492937051703357e-16,0,-1.4142135381698608,-1.4142135381698608,0,-3.6739402930577075e-16,-2,0,1.4142135381698608,-1.4142135381698608,0,0,0,5,0,0,0],"dataType":"Vec3","defaultValue":0,"length":10,"splits":{},"splitValues":[]},"texCoords":{"data":[0,0,0.125,0,0.125,0,0.25,0,0.375,0,0.5,0,0.625,0,0.75,0,0.0625,1,null,null],"dataType":"Vec2","defaultValue":1.7976931348623157e+308,"length":10,"splits":{"0":{"0":1,"7":15},"1":{"0":0,"1":3},"2":{"2":5},"3":{"3":7},"4":{"4":9},"5":{"5":11},"6":{"6":13},"7":{"7":16},"8":{"0":2,"1":4,"2":6,"3":8,"4":10,"5":12,"6":14,"7":2}},"splitValues":[{"x":0,"y":0},{"x":0.125,"y":0},{"x":0.9375,"y":1},{"x":0.25,"y":0},{"x":0.1875,"y":1},{"x":0.375,"y":0},{"x":0.3125,"y":1},{"x":0.5,"y":0},{"x":0.4375,"y":1},{"x":0.625,"y":0},{"x":0.5625,"y":1},{"x":0.75,"y":0},{"x":0.6875,"y":1},{"x":0.875,"y":0},{"x":0.8125,"y":1},{"x":0.875,"y":0},{"x":1,"y":0}]},"normals":{"data":[0.9284766912460327,0.3713906705379486,0,0.6565321683883667,0.371390700340271,0.6565321683883667,5.685280509759434e-17,0.3713906705379486,0.9284766912460327,-0.6565321683883667,0.371390700340271,0.6565321683883667,-0.9284766912460327,0.3713906705379486,1.1370561019518867e-16,-0.6565321683883667,0.371390700340271,-0.6565321683883667,-1.705583954404483e-16,0.3713906705379486,-0.9284766912460327,0.6565321683883667,0.371390700340271,-0.6565321683883667,0.8578006029129028,0.371390700340271,0.35531264543533325,0,-1,0],"dataType":"Vec3","defaultValue":1.7976931348623157e+308,"length":10,"splits":{"0":{"7":6,"8":8,"15":8},"1":{"8":9,"9":9},"2":{"9":10,"10":10},"3":{"10":11,"11":11},"4":{"11":12,"12":12},"5":{"12":13,"13":13},"6":{"13":14,"14":14},"7":{"14":15,"15":15},"8":{"1":0,"2":1,"3":2,"4":3,"5":4,"6":5,"7":7}},"splitValues":[{"x":0.35531264543533325,"y":0.371390700340271,"z":0.8578006029129028},{"x":-0.35531264543533325,"y":0.371390700340271,"z":0.8578006029129028},{"x":-0.8578006029129028,"y":0.371390700340271,"z":0.35531264543533325},{"x":-0.8578006029129028,"y":0.371390700340271,"z":-0.35531264543533325},{"x":-0.35531264543533325,"y":0.371390700340271,"z":-0.8578006029129028},{"x":0.35531264543533325,"y":0.371390700340271,"z":-0.8578006029129028},{"x":0.9284766912460327,"y":0.3713906705379486,"z":-2.2741122039037734e-16},{"x":0.8578006029129028,"y":0.371390700340271,"z":-0.35531264543533325},{"x":0,"y":-1,"z":0},{"x":0,"y":-1,"z":0},{"x":0,"y":-1,"z":0},{"x":0,"y":-1,"z":0},{"x":0,"y":-1,"z":0},{"x":0,"y":-1,"z":0},{"x":0,"y":-1,"z":0},{"x":0,"y":-1,"z":0}]}},"faceCounts":[16],"faceVertexIndices":[1,0,8,2,1,8,3,2,8,4,3,8,5,4,8,6,5,8,7,6,8,0,7,8,0,1,9,1,2,9,2,3,9,3,4,9,4,5,9,5,6,9,6,7,9,7,0,9]}'
    cone.fromJSON(JSON.parse(expectedOutput))

    const newCone = new Cone(2, 5, 8, true)
    expect(cone.toJSON()).toEqual(newCone.toJSON())
  })
})
