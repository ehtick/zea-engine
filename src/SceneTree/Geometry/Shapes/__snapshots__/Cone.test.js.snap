// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cone saves to JSON (serialization). 1`] = `
Object {
  "faceCounts": Array [
    16,
  ],
  "faceVertexIndices": Array [
    1,
    0,
    8,
    2,
    1,
    8,
    3,
    2,
    8,
    4,
    3,
    8,
    5,
    4,
    8,
    6,
    5,
    8,
    7,
    6,
    8,
    0,
    7,
    8,
    0,
    1,
    9,
    1,
    2,
    9,
    2,
    3,
    9,
    3,
    4,
    9,
    4,
    5,
    9,
    5,
    6,
    9,
    6,
    7,
    9,
    7,
    0,
    9,
  ],
  "numVertices": 10,
  "params": Object {
    "Cap": Object {
      "value": true,
    },
    "Detail": Object {
      "range": Array [
        3,
        200,
      ],
      "step": 1,
      "value": 8,
    },
    "Height": Object {
      "value": 5,
    },
    "Radius": Object {
      "value": 2,
    },
  },
  "type": "Cone",
  "vertexAttributes": Object {
    "normals": Object {
      "data": Array [
        0.9284766912460327,
        0.3713906705379486,
        0,
        0.6565321683883667,
        0.371390700340271,
        0.6565321683883667,
        5.685280509759434e-17,
        0.3713906705379486,
        0.9284766912460327,
        -0.6565321683883667,
        0.371390700340271,
        0.6565321683883667,
        -0.9284766912460327,
        0.3713906705379486,
        1.1370561019518867e-16,
        -0.6565321683883667,
        0.371390700340271,
        -0.6565321683883667,
        -1.705583954404483e-16,
        0.3713906705379486,
        -0.9284766912460327,
        0.6565321683883667,
        0.371390700340271,
        -0.6565321683883667,
        0.8578006029129028,
        0.371390700340271,
        0.35531264543533325,
        0,
        -1,
        0,
      ],
      "dataType": "Vec3",
      "defaultValue": 1.7976931348623157e+308,
      "length": 10,
      "splitValues": Array [
        Object {
          "x": 0.35531264543533325,
          "y": 0.371390700340271,
          "z": 0.8578006029129028,
        },
        Object {
          "x": -0.35531264543533325,
          "y": 0.371390700340271,
          "z": 0.8578006029129028,
        },
        Object {
          "x": -0.8578006029129028,
          "y": 0.371390700340271,
          "z": 0.35531264543533325,
        },
        Object {
          "x": -0.8578006029129028,
          "y": 0.371390700340271,
          "z": -0.35531264543533325,
        },
        Object {
          "x": -0.35531264543533325,
          "y": 0.371390700340271,
          "z": -0.8578006029129028,
        },
        Object {
          "x": 0.35531264543533325,
          "y": 0.371390700340271,
          "z": -0.8578006029129028,
        },
        Object {
          "x": 0.9284766912460327,
          "y": 0.3713906705379486,
          "z": -2.2741122039037734e-16,
        },
        Object {
          "x": 0.8578006029129028,
          "y": 0.371390700340271,
          "z": -0.35531264543533325,
        },
        Object {
          "x": 0,
          "y": -1,
          "z": 0,
        },
        Object {
          "x": 0,
          "y": -1,
          "z": 0,
        },
        Object {
          "x": 0,
          "y": -1,
          "z": 0,
        },
        Object {
          "x": 0,
          "y": -1,
          "z": 0,
        },
        Object {
          "x": 0,
          "y": -1,
          "z": 0,
        },
        Object {
          "x": 0,
          "y": -1,
          "z": 0,
        },
        Object {
          "x": 0,
          "y": -1,
          "z": 0,
        },
        Object {
          "x": 0,
          "y": -1,
          "z": 0,
        },
      ],
      "splits": Object {
        "0": Object {
          "15": 8,
          "7": 6,
          "8": 8,
        },
        "1": Object {
          "8": 9,
          "9": 9,
        },
        "2": Object {
          "10": 10,
          "9": 10,
        },
        "3": Object {
          "10": 11,
          "11": 11,
        },
        "4": Object {
          "11": 12,
          "12": 12,
        },
        "5": Object {
          "12": 13,
          "13": 13,
        },
        "6": Object {
          "13": 14,
          "14": 14,
        },
        "7": Object {
          "14": 15,
          "15": 15,
        },
        "8": Object {
          "1": 0,
          "2": 1,
          "3": 2,
          "4": 3,
          "5": 4,
          "6": 5,
          "7": 7,
        },
      },
    },
    "positions": Object {
      "data": Array [
        2,
        0,
        0,
        1.4142135381698608,
        1.4142135381698608,
        0,
        1.2246468525851679e-16,
        2,
        0,
        -1.4142135381698608,
        1.4142135381698608,
        0,
        -2,
        2.4492937051703357e-16,
        0,
        -1.4142135381698608,
        -1.4142135381698608,
        0,
        -3.6739402930577075e-16,
        -2,
        0,
        1.4142135381698608,
        -1.4142135381698608,
        0,
        0,
        0,
        5,
        0,
        0,
        0,
      ],
      "dataType": "Vec3",
      "defaultValue": 0,
      "length": 10,
      "splitValues": Array [],
      "splits": Object {},
    },
    "texCoords": Object {
      "data": Array [
        0,
        0,
        0.125,
        0,
        0.125,
        0,
        0.25,
        0,
        0.375,
        0,
        0.5,
        0,
        0.625,
        0,
        0.75,
        0,
        0.0625,
        1,
        Infinity,
        Infinity,
      ],
      "dataType": "Vec2",
      "defaultValue": 1.7976931348623157e+308,
      "length": 10,
      "splitValues": Array [
        Object {
          "x": 0,
          "y": 0,
        },
        Object {
          "x": 0.125,
          "y": 0,
        },
        Object {
          "x": 0.9375,
          "y": 1,
        },
        Object {
          "x": 0.25,
          "y": 0,
        },
        Object {
          "x": 0.1875,
          "y": 1,
        },
        Object {
          "x": 0.375,
          "y": 0,
        },
        Object {
          "x": 0.3125,
          "y": 1,
        },
        Object {
          "x": 0.5,
          "y": 0,
        },
        Object {
          "x": 0.4375,
          "y": 1,
        },
        Object {
          "x": 0.625,
          "y": 0,
        },
        Object {
          "x": 0.5625,
          "y": 1,
        },
        Object {
          "x": 0.75,
          "y": 0,
        },
        Object {
          "x": 0.6875,
          "y": 1,
        },
        Object {
          "x": 0.875,
          "y": 0,
        },
        Object {
          "x": 0.8125,
          "y": 1,
        },
        Object {
          "x": 0.875,
          "y": 0,
        },
        Object {
          "x": 1,
          "y": 0,
        },
      ],
      "splits": Object {
        "0": Object {
          "0": 1,
          "7": 15,
        },
        "1": Object {
          "0": 0,
          "1": 3,
        },
        "2": Object {
          "2": 5,
        },
        "3": Object {
          "3": 7,
        },
        "4": Object {
          "4": 9,
        },
        "5": Object {
          "5": 11,
        },
        "6": Object {
          "6": 13,
        },
        "7": Object {
          "7": 16,
        },
        "8": Object {
          "0": 2,
          "1": 4,
          "2": 6,
          "3": 8,
          "4": 10,
          "5": 12,
          "6": 14,
          "7": 2,
        },
      },
    },
  },
}
`;
